graph2tac/loader/py_data_server.py

```diff
@@ -171,23 +171,67 @@ def cluster_to_graph(self, cluster):
             ], dtype = self._def_name_dtype)
         )
 
-    
+class Splitter:
+    def datapoint(self, d : Definition):
+        raise Exception("Not implemented")
+    def cluster(self, d : list[Definition]):
+        raise Exception("Not implemented")
+
+class SplitByHash(Splitter):
+    def __init__(self, split = (8,1,1), random_seed = 0):
+        self.data_server = None # to be set by data_server
+        self.split = split
+        self.random_seed = random_seed
+    def datapoint(self, d : Definition):
+        # to make it identical to vasily's loader
+        ident_64 = np.array(d.node.identity, dtype = np.uint64).item()
+        return get_split_label(ident_64, self.split, self.random_seed)
+    def cluster(self, d : list[Definition]):
+        return 0
+
+class SplitByFilePrefix(Splitter):
+    def __init__(self, *prefixes_per_label):
+        self.data_server = None # to be set by data_server
+        self.prefixes_per_label = []
+        for label, prefixes in enumerate(prefixes_per_label):
+            if isinstance(prefixes, str):
+                prefixes = [prefixes]
+            else:
+                prefixes = list(prefixes)
+            self.prefixes_per_label.append((label+1, prefixes))
+    def datapoint(self, d : Definition):
+        fname = self.data_server.graphid_to_fname[d.node.graph]
+        for label, prefixes in reversed(self.prefixes_per_label):
+            if any(fname.startswith(prefix) for prefix in prefixes):
+                return label
+        return 0
+    def cluster(self, d : list[Definition]):
+        return self.datapoint(d[0])
+
 class DataServer(AbstractDataServer):
     def __init__(self,
                  data_dir: Path,
                  max_subgraph_size,
-                 split: tuple[int, int, int] = (8, 1, 1),
+                 split = (8, 1, 1),
                  bfs_option = True,
                  split_random_seed = 0,
                  restrict_to_spine: bool = False,
                  stop_at_definitions: bool = True,
     ):
         super().__init__(max_subgraph_size, bfs_option, stop_at_definitions)
         self.data_dir = data_dir
-        self.split = split
-        self.split_random_seed = split_random_seed
         self.restrict_to_spine = restrict_to_spine
 
+        if isinstance(split, tuple) and isinstance(split[0], int):
+            self.split = SplitByHash(split, split_random_seed)
+        elif isinstance(split, tuple) and isinstance(split[0], (str, list)) and isinstance(split[0][0], str):
+            self.split = SplitByFilePrefix(split)
+        elif isinstance(split, Splitter):
+            self.split = split # ignore split_random_seed
+        else:
+            raise Exception("Cannot interpret split:", split)
+        self.split.data_server = self
```

Comment:
Passing directly a Splitter is among the options. I agree I am sort of lazy / hacky here with keeping backward compatibility. At some point, we need to parse the config file, and create the appropriate Splitter. Now, I am doing it here but I agree there could be a better option.

Reply:
For example, TF2 is using the old split argument, and I am not sure if we are already fully getting rid of it.

Reply:
Yes, we are getting rid of TF2 (Jason should be working on it). I think that the time for keeping backwards compatibility is over. I would just change it to what we think is good code.

Reply:
See my suggestion below.  And yes, I'm getting rid of TF2 this week hopefully.

Reply:
Actually, I'll just repeat it.  We can use this pattern often used by Fidel in the yaml files:
```
split_method: hash
split:
  proportions:
  - 0
  - 1
  split_random_seed: 0
```
So then we pass that to the DataServer as `split_method: str` and `split: dict`.  This will make it easy to add more split methods as needed.